onnxscript/rewriter/ort_fusions/attention.py

```diff
@@ -29,32 +33,43 @@ def pattern(
         qkv_bias,
         # mask_index,
         past,
-        # attention_bias,
+        attention_bias,
         num_heads,
         # scale,
+        q_mul,
+        k_mul,
+        v_mul,
     ):
-        projected = op.MatMul(input, qkv_weight)
-        # Add bias if present
-        if self._has_input_bias:
-            projected = op.Add(projected, qkv_bias)
+        if self._no_slice:
+            query_BSD = op.MatMul(input, q_mul)
+            key_BSD = op.MatMul(input, k_mul)
+            value_BSD = op.MatMul(input, v_mul)
+        else:
+            projected = op.MatMul(input, qkv_weight)
 
-        # Slice packed Matmul QKV into Q, K, and V
-        # Q, K, and V are of shape (B, S, D)
-        query_BSD = op.Slice(
-            projected,
-            _allow_other_inputs=True,
-            _outputs=["query_mm_sliced"],
-        )
-        key_BSD = op.Slice(
-            projected,
-            _allow_other_inputs=True,
-            _outputs=["key_mm_sliced"],
-        )
-        value_BSD = op.Slice(
-            projected,
-            _allow_other_inputs=True,
-            _outputs=["value_mm_sliced"],
-        )
+            # Slice packed Matmul QKV into Q, K, and V
+            # Q, K, and V are of shape (B, S, D)
+            query_BSD = op.Slice(
+                projected,
+                pattern.ANY_VALUE,  # starts
```

Comment:
Are there checks down below to ensure that the starts/ends have the right values? I couldn't easily locate them. But it is a big rule, so I may have missed it.

Reply:
Added